core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -531,6 +509,60 @@ class TeamPolicyInternal<Kokkos::Cuda, Properties...>
     return *this;
   }
 #endif
+
+  template <class FunctorType, class ClosureType>
+  int internal_team_size_max(const FunctorType& f) const {
+    using closure_type = ClosureType;
+    typedef Impl::FunctorValueTraits<FunctorType, typename traits::work_tag>
+        functor_value_traits;
+
+    cudaFuncAttributes attr =
+        CudaParallelLaunch<closure_type, typename traits::launch_bounds>::
+            get_cuda_func_attributes();
+    const int block_size =
+        Kokkos::Impl::cuda_get_max_block_size<FunctorType,
+                                              typename traits::launch_bounds>(
+            space().impl_internal_space_instance(), attr, f,
+            (size_t)vector_length(),
+            (size_t)team_scratch_size(0) + 2 * sizeof(double),
+            (size_t)thread_scratch_size(0) + sizeof(double) +
+                ((functor_value_traits::StaticValueSize != 0)
+                     ? 0
+                     : functor_value_traits::value_size(f)));
+
+    // Currently we require Power-of-2 team size for reductions.
+    int p2 = 1;
+    while (p2 <= block_size) p2 *= 2;
+    p2 /= 2;
+    return p2 / vector_length();
+  }
+
+  template <class FunctorType, class ClosureType>
+  int internal_team_size_recommended(const FunctorType& f) const {
+    using closure_type = ClosureType;
+    typedef Impl::FunctorValueTraits<FunctorType, typename traits::work_tag>
+        functor_value_traits;
+
+    cudaFuncAttributes attr =
+        CudaParallelLaunch<closure_type, typename traits::launch_bounds>::
+            get_cuda_func_attributes();
+    const int block_size =
+        Kokkos::Impl::cuda_get_opt_block_size<FunctorType,
+                                              typename traits::launch_bounds>(
+            space().impl_internal_space_instance(), attr, f,
+            (size_t)vector_length(),
+            (size_t)team_scratch_size(0) + 2 * sizeof(double),
+            (size_t)thread_scratch_size(0) + sizeof(double) +
+                ((functor_value_traits::StaticValueSize != 0)
+                     ? 0
+                     : functor_value_traits::value_size(f)));
+
+    // Currently we require Power-of-2 team size for reductions.
+    int p2 = 1;
+    while (p2 <= block_size) p2 *= 2;
+    p2 /= 2;
+    return p2 / vector_length();
+  }
```

Comment:
There is **far** to much copy-pasted code in this function. AFAICT literally only line 550 is different. Please restructure to avoid copy-pasta. If there's absolutely no way to do that, please provide a long, detailed comment justifying the copy-paste, detailing *exactly* the differences between the two functions (e.g., *each* and every copied line should be commented with "copied from above because...", and each line where copies from the original can't be avoided should be commented with "copied below because..."), and explaining why there's no way to avoid introducing an extra copy of application logic into the codebase. (Be warned, though, that my bar for approving such things is very high at this point. In my assessment, a large portion of the technical debt in Kokkos currently comes from this kind of thing.)

Reply:
I think you're right, and I can refactor this to instead be one function which takes in an argument which could be the different things called on those lines, if that's an acceptable refactoring

Reply:
Great :-) Exactly what I meant. If it gets too complicated, that's fine, it's just that my bar for "too complicated" is at least that it has to be worth commenting every line that was copied in both places and explaining why